Add experimental provenance support and versioning

bruth opened the pull request:
The README has been updated with the new endpoints.

Signed-off-by: Byron Ruth b@devel.io

bruth committed:
Add experimental provenance support and versioning

The README has been updated with the new endpoints.

Signed-off-by: Byron Ruth <b@devel.io>

bruth commented:
@aaron0browne @l-k- The `/pedsnet` and `/i2b2` endpoints now support the `asof` and `ref` parameters to get the state of the document at some point in time. Accompanied with this are the provenance endpoints that are still experimental..

bruth commented:
The one bit missing is a `version` parameter, e.g. `/pedsnet?version=2.0.0`. @aaron0browne do you suggest a strategy for mapping a data model version to a particular commit of the document?

gracecodes144 commented:
Evanette is in process of creating tags for the repo that map to versions of the convention document. Of course, that is just the PEDSnet conventions document, and I hadn't thought about the fact that the i2b2 conventions document lives in the same repo...

In any case, so far the idea is that the tags will look like v2.1_1 where v2.1 is the data model version and 1 is the serial version of the conventions document, since the two are free to change independently of each other.

bruth commented:
An alternative is to hard code them into the service itself (or have a config for it) that maps version to commit. The latest version can be specified as well, but is not bound since new commits will up it. This would handle the two documents per repo.

gracecodes144 commented:
The tags are intended to be useful for site ETL users as well. Maybe we just need to add a data model prefix to the tag, like pedsnet_v2.1_1?

bruth commented:
Is the idea that ETL users will look on the releases page?

gracecodes144 commented:
Yup

bruth commented:
Ok, then your suggested tag names make sense.

l-k- commented on services/etl/provenance.py:
```diff
@@ -0,0 +1,279 @@
+
+- `domain` - The top-level domain the entity should be written to.
+- `ident` - The unique identity value of the entity. Entities that do
+not (or should not) be referenced by other entities do require this
```
Is this correct, or should this be "do not require this to be set"? Just wanted to make sure.
Looks good otherwise.

bruth replied:
Thanks. It should state why it should be set rather than why not.

> An identity should be provided if the entity is a continuant in the domain so all occurrents can be associated.

bruth committed:
Modify format of provenance output

Previously the `domain` could have been used to delegate which domain
the fact should be written to or the domain the entity is associated with.
Since this was ambiguous, the consumer is now responsible to delegating
where the fact should be written. The `domain` and `name` now identify
the entity.

References to other entities has been broken out `attrs` as `refs` to make
it easier for consumers to parse. Further, it will remove ambiguity for
nested documents in `attrs`.

A `batch` attribute has been defined which is used for tagging all entities
within a batch or snapshot with the same identifier. For stream processing
applications, this provides additional context of which entity came from
where.

In addition to the above, several bugs were fixed related to bad references.

Signed-off-by: Byron Ruth <b@devel.io>

bruth committed:
Reorganize various functions for reuse

Add cache for fetched content at a specific commit.

Signed-off-by: Byron Ruth <b@devel.io>

bruth committed:
Add endpoints for provenance changes

Signed-off-by: Byron Ruth <b@devel.io>

bruth committed:
Add support for versions

There is no method for deriving the data model version from the
document, so a hard-coded set of versions have been added to their
most recent known commit. The latest version points to ‘master’ so
it will remain at the latest commit.

The other changes involved removing redundant commit parsing.

Signed-off-by: Byron Ruth <b@devel.io>

bruth committed:
Increase request timeout to GitHub

Signed-off-by: Byron Ruth <b@devel.io>
